output.c

```diff
@@ -75,6 +100,46 @@ output_disable(struct cg_output *output)
 	wlr_output_enable(wlr_output, false);
 	wlr_output_layout_remove(output->server->output_layout, wlr_output);
 	wlr_output_commit(wlr_output);
+
+	update_output_manager_config(output->server);
+}
+
+static bool
+output_apply_config(struct cg_output *output, struct wlr_output_configuration_head_v1 *head, bool test_only)
+{
+	wlr_output_enable(output->wlr_output, head->state.enabled);
+
+	if (head->state.enabled) {
+		/* Do not mess with these parameters for output to be disabled */
+		wlr_output_set_scale(output->wlr_output, head->state.scale);
+		wlr_output_set_transform(output->wlr_output, head->state.transform);
+
+		if (head->state.mode) {
+			wlr_output_set_mode(output->wlr_output, head->state.mode);
+		} else {
+			wlr_output_set_custom_mode(output->wlr_output, head->state.custom_mode.width,
+						   head->state.custom_mode.height, head->state.custom_mode.refresh);
+		}
+	}
+
+	if (test_only) {
+		bool ret = wlr_output_test(output->wlr_output);
+		wlr_output_rollback(output->wlr_output);
+		return ret;
+	}
+
+	/* Apply output configuration */
+	if (head->state.enabled) {
+		wlr_output_layout_add(output->server->output_layout, head->state.output, head->state.x, head->state.y);
```

Comment:
We should add the output to the layout only after `wlr_output_commit()` succeeds. If the commit fails, we don't want to apply the rest of the configuration changes.